sympy/integrals/transforms.py

```diff
@@ -1186,499 +1194,360 @@ def _laplace_build_rules(t, s):
+    masking property of the delta distribution. For example, if it gets
+    ``(DiracDelta(t-a)*f(t), t, s)``, it will return
+    ``(f(a)*exp(-a*s), -a, True)``.
+    """
+    # This rule is not in Bateman54
+    f, p, c = _laplace_ct(f)
+
+    a = Wild('a', exclude=[t])
+    b = Wild('b', exclude=[t])
+
+    y = Wild('y')
+    z = Wild('z')
+    ma1 = f.match(DiracDelta(y)*z)
+    if ma1 and not ma1[z].has(DiracDelta):
+        ma2 = ma1[y].collect(t).match(b*t-a)
+        if ma2:
+            debug('_laplace_apply_prog_rules match:')
+            debug('      f:    %s ( %s, %s )'%(f, ma1, ma2))
+            debug('      rule: multiply with DiracDelta')
+            loc = ma2[a]/ma2[b]
+            if re(loc)>=0 and im(loc)==0:
+                r = exp(-ma2[a]/ma2[b]*s)*ma1[z].subs(t, ma2[a]/ma2[b])/ma2[b]
+                return True, (r, p, c)
             else:
-                return L
-    return None
+                return True, _laplace_ct(0)
+    return False, (f, p, c)
 
 def _laplace_rule_trig(f, t, s, doit=True, **hints):
     """
-    This internal helper function tries to transform a product containing a
-    trigonometric function (`sin`, `cos`, `sinh`, `cosh`, ) and returns
-    `None` if it cannot do it.
+    This function covers trigonometric factors. All of the rules have a
+    similar form: ``trig(y)*z`` is matched, and then two copies of the Laplace
+    transform of `z` are shifted in the s Domain and added with a weight.
+
+    The parameters in the tuples are (fm, nu, s1, s2, sd):
+      fm: Function to match
+      nu: Number of the rule, for debug purposes
+      s1: weight of the sum, 'I' for sin and '1' for all others
+      s2: sign of the second copy of the Laplace transform of z
+      sd: shift direction; shift along real or imaginary axis if `1` or `I`
+
+    The convergence plane is changed only if a frequency shift is done along
+    the real axis.
     """
-    _simplify = hints.pop('simplify', True)
+    # These rules follow from Bateman54, 4.1.5 and Euler's formulas
+    f, p, c = _laplace_ct(f)
+
     a = Wild('a', exclude=[t])
     y = Wild('y')
     z = Wild('z')
-    # All of the rules have a very similar form: trig(y)*z is matched, and then
-    # two copies of the Laplace transform of z are shifted in the s Domain
-    # and added with a weight; see rules 1.6 to 1.9 in
-    # http://eqworld.ipmnet.ru/en/auxiliary/inttrans/laplace1.pdf
-    # The parameters in the tuples are (fm, nu, s1, s2, sd):
-    #   fm: Function to match
-    #   nu: Number of the rule, for debug purposes
-    #   s1: weight of the sum, 'I' for sin and '1' for all others
-    #   s2: sign of the second copy of the Laplace transform of z
-    #   sd: shift direction; shift along real or imaginary axis if `1` or `I`
     trigrules = [(sinh(y), '1.6',  1, -1, 1), (cosh(y), '1.7', 1, 1, 1),
                  (sin(y),  '1.8', -I, -1, I), (cos(y),  '1.9', 1, 1, I)]
```

Comment:
In a power like `sin(t)**10 * f(t)` this will extract only one factor of `sin(t)` at a time which in efficient. It should be possible to derive the more general rule for powers e.g.:
```python
In [23]: st = lambda f: (-I*f.subs(s, s-I) - f.subs(s, s+I))/2

In [24]: F = Function('F')(s)

In [25]: st(F)
Out[25]: 
  ⅈ⋅F(s - ⅈ)   F(s + ⅈ)
- ────────── - ────────
      2           2    

In [26]: st(st(F))
Out[26]: 
    ⎛  F(s)   ⅈ⋅F(s - 2⋅ⅈ)⎞                      
  ⅈ⋅⎜- ──── - ────────────⎟                      
    ⎝   2          2      ⎠   ⅈ⋅F(s)   F(s + 2⋅ⅈ)
- ───────────────────────── + ────── + ──────────
              2                 4          4     

In [27]: st(st(F)).simplify()
Out[27]: 
ⅈ⋅F(s)   F(s - 2⋅ⅈ)   F(s + 2⋅ⅈ)
────── - ────────── + ──────────
  2          4            4     

In [28]: st(st(st(F))).simplify()
Out[28]: 
ⅈ⋅F(s - 3⋅ⅈ)   3⋅F(s - ⅈ)   3⋅ⅈ⋅F(s + ⅈ)   F(s + 3⋅ⅈ)
──────────── + ────────── - ──────────── - ──────────
     8             8             8             8  
```
The coefficients follow a pattern with binomial coefficients:
```
In [39]: factor_terms(expand((-(x + I)/2)**3))
Out[39]: 
 3        2          
x  + 3⋅ⅈ⋅x  - 3⋅x - ⅈ
─────────────────────
          8 
```
The function `st` generates a complexified version of the Pascal's triangle. For `cosh` it actually is the triangle.

Reply:
Also why doesn't the second example work:
```
In [86]: laplace_transform(cosh(t)*f(t), t, s)
Out[86]: 
⎛LaplaceTransform(f(t), t, s - 1)   LaplaceTransform(f(t), t, s + 1)          ⎞
⎜──────────────────────────────── + ────────────────────────────────, -∞, True⎟
⎝               2                                  2                          ⎠

In [87]: laplace_transform(cosh(t)**2*f(t), t, s)
Out[87]: 
⎛                ⎛         2         ⎞          ⎞
⎝LaplaceTransform⎝f(t)⋅cosh (t), t, s⎠, -∞, True⎠

```